File: android/pairing-core/src/main/kotlin/org/phonebridge/pairing/PairingFrames.kt
package org.phonebridge.pairing

class PairingProtocolException : Exception("pairing_protocol_error")

object PairingFrames {
    const val HEADER_SIZE = 9
    const val MAX_PAYLOAD = 8192
    const val MAX_SESSION_BYTES = 32768
    internal val order = intArrayOf(1, 2, 0x11, 0x12, 0x21, 0x22, 0x31, 0x32)

    internal fun length(header: ByteArray, expected: Int): Int {
        if (header.size != HEADER_SIZE || !header.copyOfRange(0,4).contentEquals(byteArrayOf(80,66,80,49)) ||
            header[4].toInt() and 255 != expected) throw PairingProtocolException()
        var size = 0L
        for (i in 5..8) size = (size shl 8) or (header[i].toLong() and 255)
        val valid = when (expected) {
            1 -> size == 33L
            2 -> size in 54L..4149L
            0x11, 0x12 -> size == 1600L
            0x21, 0x22 -> size == 800L
            0x31, 0x32 -> size == 32L
            else -> false
        }
        if (!valid || size > MAX_PAYLOAD) throw PairingProtocolException()
        return size.toInt()
    }

    fun validate(frame: ByteArray, expected: Int) {
        if (frame.size < HEADER_SIZE || length(frame.copyOfRange(0,HEADER_SIZE),expected) != frame.size-HEADER_SIZE)
            throw PairingProtocolException()
    }

    internal fun encode(type: Int, body: ByteArray): ByteArray {
        val out = ByteArray(HEADER_SIZE+body.size)
        byteArrayOf(80,66,80,49,type.toByte()).copyInto(out)
        for (i in 0..3) out[5+i] = (body.size ushr (24-8*i)).toByte()
        body.copyInto(out,HEADER_SIZE); validate(out,type); return out
    }
}

/** One expected frame. The transport must check EOF and reject surplus bytes. */
class FrameAccumulator(private val expected: Int) {
    private val header = ByteArray(PairingFrames.HEADER_SIZE)
    private var frame: ByteArray? = null
    private var count = 0
    private var failed = false
    val isComplete: Boolean get() = !failed && frame != null && count == frame!!.size
    init { if (!PairingFrames.order.contains(expected)) throw PairingProtocolException() }

    fun feed(bytes: ByteArray) {
        try {
            if (failed || isComplete) throw PairingProtocolException()
            var offset = 0
            if (count < header.size) {
                val n = minOf(bytes.size,header.size-count)
                bytes.copyInto(header,count,0,n); count += n; offset += n
                if (count < header.size) return
                frame = ByteArray(header.size+PairingFrames.length(header,expected))
                header.copyInto(frame!!)
            }
            if (bytes.size-offset > frame!!.size-count) throw PairingProtocolException()
            bytes.copyInto(frame!!,count,offset); count += bytes.size-offset
        } catch (_: Exception) { failed = true; frame = null; throw PairingProtocolException() }
    }

    fun endOfInput() {
        if (!isComplete) { failed = true; frame = null; throw PairingProtocolException() }
    }
    fun getFrame(): ByteArray { endOfInput(); return frame!!.copyOf() }
}

File: windows/src/PhoneBridge.Pairing/PairingFrames.cs
using System.Buffers.Binary;

namespace PhoneBridge.Pairing;

public sealed class PairingProtocolException : Exception
{
    public PairingProtocolException() : base("pairing_protocol_error") { }
}

public static class PairingFrames
{
    public const int HeaderSize = 9;
    public const int MaxPayload = 8192;
    public const int MaxSessionBytes = 32768;
    internal static readonly byte[] Order = [1, 2, 0x11, 0x12, 0x21, 0x22, 0x31, 0x32];

    internal static int Length(ReadOnlySpan<byte> header, byte expectedType)
    {
        if (header.Length != HeaderSize || !header[..4].SequenceEqual("PBP1"u8) || header[4] != expectedType)
            throw new PairingProtocolException();
        uint size = BinaryPrimitives.ReadUInt32BigEndian(header[5..]);
        bool valid = expectedType switch
        {
            1 => size == 33,
            2 => size is >= 54 and <= 4149,
            0x11 or 0x12 => size == 1600,
            0x21 or 0x22 => size == 800,
            0x31 or 0x32 => size == 32,
            _ => false
        };
        if (!valid || size > MaxPayload) throw new PairingProtocolException();
        return (int)size;
    }

    public static void Validate(ReadOnlySpan<byte> frame, byte expectedType)
    {
        if (frame.Length < HeaderSize || Length(frame[..HeaderSize], expectedType) != frame.Length - HeaderSize)
            throw new PairingProtocolException();
    }

    internal static byte[] Encode(byte type, byte[] payload)
    {
        var frame = new byte[HeaderSize + payload.Length];
        "PBP1"u8.CopyTo(frame); frame[4] = type;
        BinaryPrimitives.WriteUInt32BigEndian(frame.AsSpan(5), (uint)payload.Length);
        payload.CopyTo(frame, HeaderSize); Validate(frame, type); return frame;
    }
}

/// <summary>One expected frame only. A transport must call EndOfInput on EOF and reject surplus bytes.</summary>
public sealed class FrameAccumulator
{
    private readonly byte _expected;
    private readonly byte[] _header = new byte[PairingFrames.HeaderSize];
    private byte[]? _frame;
    private int _count;
    private bool _failed;
    public bool IsComplete => !_failed && _frame is not null && _count == _frame.Length;

    public FrameAccumulator(byte expectedType)
    {
        if (!PairingFrames.Order.Contains(expectedType)) throw new PairingProtocolException();
        _expected = expectedType;
    }

    public void Feed(ReadOnlySpan<byte> bytes)
    {
        try
        {
            if (_failed || IsComplete) throw new PairingProtocolException();
            if (_count < _header.Length)
            {
                int n = Math.Min(bytes.Length, _header.Length - _count);
                bytes[..n].CopyTo(_header.AsSpan(_count)); _count += n; bytes = bytes[n..];
                if (_count < _header.Length) return;
                _frame = new byte[_header.Length + PairingFrames.Length(_header, _expected)];
                _header.CopyTo(_frame, 0);
            }
            if (bytes.Length > _frame!.Length - _count) throw new PairingProtocolException();
            bytes.CopyTo(_frame.AsSpan(_count)); _count += bytes.Length;
        }
        catch { _failed = true; _frame = null; throw new PairingProtocolException(); }
    }

    public void EndOfInput()
    {
        if (!IsComplete) { _failed = true; _frame = null; throw new PairingProtocolException(); }
    }

    public byte[] GetFrame() { EndOfInput(); return (byte[])_frame!.Clone(); }
}
